akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed.cluster
+
+import akka.actor.Address
+import akka.annotation.DoNotInherit
+import akka.cluster.ClusterEvent.{ ClusterDomainEvent, CurrentClusterState, InitialStateAsSnapshot, SubscriptionInitialStateMode }
+import akka.cluster._
+import akka.typed.internal.adapter.ActorSystemAdapter
+import akka.typed.{ ActorRef, ActorSystem, Extension, ExtensionId }
+
+import scala.collection.immutable
+import scala.reflect.ClassTag
+
+object Cluster extends ExtensionId[Cluster] {
+
+  sealed trait ClusterStateSubscription
+  // we can't have initial cluster as snapshot as that wouldn't be a ClusterDomainEvent?
```

Comment:
I think we must have that, to avoid race conditions between sending the snapshot and events occurring after the snaphot.

We could add a new common interface, such as `ClusterMembershipInfo`, but that would be very wide if one are only interested in `MemberUp`.

We could take two `ActorRef` parameters, `ActorRef[CurrentClusterState]` and `ActorRef[T]`. The subscriber actor would then have to use the function adapters to convert.

Reply:
or we only support the `InitialStateAsEvents` mode,

if one really want the snapshot mode it would be possible with this (rather complicated, and inefficient) approach:
* cluster. subscriptions ! Subscribe(...)
* cluster. subscriptions ! GetCurrentState(...)
* discard all ClusterDomainEvent messages until CurrentClusterState is received
* receive CurrentClusterState
* receive new ClusterDomainEvent